README.md
# jQuery hashchange event #
[http://benalman.com/projects/jquery-hashchange-plugin/](http://benalman.com/projects/jquery-hashchange-plugin/)

Version: 2.0, Last updated: 4/18/2020

This jQuery plugin enables very basic bookmarkable #hash history via a cross-browser window.onhashchange event.

Visit the [project page](http://benalman.com/projects/jquery-hashchange-plugin/) for more information and usage examples!


## Documentation ##
[http://benalman.com/code/projects/jquery-hashchange/docs/](http://benalman.com/code/projects/jquery-hashchange/docs/)


## Examples ##
These working examples, complete with fully commented code, illustrate a few
ways in which this plugin can be used.

[http://benalman.com/code/projects/jquery-hashchange/examples/hashchange/](http://benalman.com/code/projects/jquery-hashchange/examples/hashchange/)
[http://benalman.com/code/projects/jquery-hashchange/examples/document_domain/](http://benalman.com/code/projects/jquery-hashchange/examples/document_domain/)

## Support and Testing ##
Information about what version or versions of jQuery this plugin has been
tested with, what browsers it has been tested in, and where the unit tests
reside (so you can test it yourself).

### jQuery Versions ###
1.2.6, 1.3.2, 1.4.1, 1.4.2

### Browsers Tested ###
Internet Explorer 6-8, Firefox 2-4, Chrome 5-6, Safari 3.2-5, Opera 9.6-10.60, iPhone 3.1, Android 1.6-2.2, BlackBerry 4.6-5.

### Unit Tests ###
[http://benalman.com/code/projects/jquery-hashchange/unit/](http://benalman.com/code/projects/jquery-hashchange/unit/)


## A more robust solution ##

This plugin is, by design, very basic. If you want to add lot of extra utility around getting and setting the hash as a state, and parsing and merging fragment params, check out the [jQuery BBQ](http://benalman.com/projects/jquery-bbq-plugin/) plugin. It includes this plugin at its core, plus a whole lot more, and has thorough documentation and examples as well. You can't have too much of a good thing!


## Known issues ##

While this jQuery hashchange event implementation is quite stable and robust, there are a few unfortunate browser bugs surrounding expected hashchange event-based behaviors, independent of any JavaScript window.onhashchange abstraction. See the following examples for more information:

Chrome: Back Button
[http://benalman.com/code/projects/jquery-hashchange/examples/bug-chrome-back-button/](http://benalman.com/code/projects/jquery-hashchange/examples/bug-chrome-back-button/)

Firefox: Remote XMLHttpRequest
[http://benalman.com/code/projects/jquery-hashchange/examples/bug-firefox-remote-xhr/](http://benalman.com/code/projects/jquery-hashchange/examples/bug-firefox-remote-xhr/)

WebKit: Back Button in an Iframe
[http://benalman.com/code/projects/jquery-hashchange/examples/bug-webkit-hash-iframe/](http://benalman.com/code/projects/jquery-hashchange/examples/bug-webkit-hash-iframe/)

Safari: Back Button from a different domain
[http://benalman.com/code/projects/jquery-hashchange/examples/bug-safari-back-from-diff-domain/](http://benalman.com/code/projects/jquery-hashchange/examples/bug-safari-back-from-diff-domain/)

Also note that should a browser natively support the window.onhashchange
event, but not report that it does, the fallback polling loop will be used.


## Release History ##

[CHANGELOG.md](CHANGELOG.md)


## License ##
Copyright (c) 2010 "Cowboy" Ben Alman
Dual licensed under the MIT and GPL licenses.
[http://benalman.com/about/license/](http://benalman.com/about/license/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0] - 2020-04-18

## [1.4] - 2020-04-18

### Fixed
- Support for jQuery 1.9
- IE detection

### Added
- Unit test for jQuery 1.7.2
- package.json
- publish to NPM

## [1.3] - 2010-07-21
- Reorganized IE6/7 Iframe code to make it more
  "removable" for mobile-only development. Added IE6/7 document.title
  support. Attempted to make Iframe as hidden as possible by using
  techniques from http://www.paciellogroup.com/blog/?p=604. Added 
  support for the "shortcut" format $(window).hashchange( fn ) and
  $(window).hashchange() like jQuery provides for built-in events.
  Renamed jQuery.hashchangeDelay to <jQuery.fn.hashchange.delay> and
  lowered its default value to 50. Added <jQuery.fn.hashchange.domain>
  and <jQuery.fn.hashchange.src> properties plus document-domain.html
  file to address access denied issues when setting document.domain in
  IE6/7.

## [1.2] - 2010-01-11
- Fixed a bug where coming back to a page using this plugin
  from a page on another domain would cause an error in Safari 4. Also,
  IE6/7 Iframe is now inserted after the body (this actually works),
  which prevents the page from scrolling when the event is first bound.
  Event can also now be bound before DOM ready, but it won't be usable
  before then in IE6/7.

## [1.1] - 2010-01-21
- Incorporated document.documentMode test to fix IE8 bug
  where browser version is incorrectly reported as 8.0, despite
  inclusion of the X-UA-Compatible IE=EmulateIE7 meta tag.

## [1.0] - 2010-01-09
- Initial Release. Broke out the jQuery BBQ event.special
  window.onhashchange functionality into a separate plugin for users
  who want just the basic event & back button support, without all the
  extra awesomeness that BBQ provides. This plugin will be included as
  part of jQuery BBQ, but also be available separately.

[Unreleased]: https://github.com/cowboy/jquery-hashchange/compare/v1.3...HEAD
[v1.3]: https://github.com/cowboy/jquery-hashchange/compare/v1.2...v1.3
[v1.2]: https://github.com/cowboy/jquery-hashchange/compare/v1.1...v1.2
[v1.1]: https://github.com/cowboy/jquery-hashchange/compare/v1.0...v1.1
[v1.0]: https://github.com/cowboy/jquery-hashchange/releases/tag/v1.0


unit/unit.js
// Not sure why this isn't set by default in qunit.js..
QUnit.jsDump.HTML = false;

(function ($) { // START CLOSURE

  var back_broken = /(?:chrome|blackberry)/i.test(navigator.userAgent),
    aps = Array.prototype.slice,
    opt = Object.prototype.toString;

  function notice(txt) {
    if (txt) {
      $('#notice').html(txt);
    } else {
      $('#notice').hide();
    }
  };

  function run_many_tests() {
    var tests = aps.call(arguments),
      delay = typeof tests[0] === 'number' && tests.shift(),
      func_each = $.isFunction(tests[0]) && tests.shift(),
      func_done = $.isFunction(tests[0]) && tests.shift(),
      result;

    function set_result(i, test) {
      result = opt.call(test) === '[object Array]' // 1.2.6 didn't have $.isArray()
        ? func_each.apply(this, test)
        : $.isFunction(test)
          ? test(result)
          : '';
    };

    function next() {
      test && test.func && test.func(result);
      if (tests.length) {
        set_result(0, tests.shift());
        setTimeout(next, delay);
      } else {
        func_done && func_done();
        start();
      }
    };

    if (delay) {
      QUnit.stop(); // BlackBerry 5 explodes if "stop()" is called by itself.
      next();

    } else {
      $.each(tests, set_result);
      func_done && func_done();
    }
  }

  ////////////////////////////////////////////////////////////////////////////////
  ////////////////////////////////////////////////////////////////////////////////

  var hash;

  // Pulled out of the plugin, pretty much.
  function get_fragment(url) {
    url = url || location.href;
    return url.replace(/^[^#]*#?(.*)$/, '$1');
  };

  if (window.document_domain_set) {
    $.fn.hashchange.src = '../document-domain.html';
    $.fn.hashchange.domain = document.domain;
  }

  // Event can be bound before DOM ready.
  $(window).bind('hashchange', function (e) {
    hash = get_fragment();
  });

  $(function () {
    $('#jq_version').html($.fn.jquery);

    module('jQuery hashchange event');

    test('window.onhashchange', function () {
      expect(8);

      var expected_hash, arr = [], msg = 'Testing window.onhashchange and history';

      location.hash = '#a';
      equals(get_fragment(), 'a', 'hash should be set properly');

      $(window).trigger('hashchange');
      equals(hash, 'a', 'hashchange triggered manually (long form)');

      hash = false;
      $(window).hashchange();
      equals(hash, 'a', 'hashchange triggered manually (short form)');

      run_many_tests(
        // run asynchronously
        250,

        // execute this for each array item
        function (new_hash) {
          notice(msg += '.');
          expected_hash = new_hash;
          location.hash = '#' + new_hash;
        },

        // execute this at the end
        function () {
          notice();
        },

        // tests:

        function (result) {
          // pause at the start, since the hash was just set to '#a'.
        },

        ['b'],

        function (result) {
          equals(hash, expected_hash, 'hash should be set properly');
        },

        [''],

        function (result) {
          equals(hash, expected_hash, 'hash should be set properly');
        },

        ['d'],

        function (result) {
          equals(hash, expected_hash, 'hash should be set properly');
        },

        function (result) {
          $(window).unbind('hashchange');
        },

        function (result) {
          $(window).hashchange(function (evt) {
            arr.push(get_fragment());
          });
        },

        function (result) {
          !back_broken && history.go(-1);
        },

        function (result) {
          !back_broken && history.go(-1);
        },

        function (result) {
          !back_broken && history.go(-1);
        },

        function (result) {
          if (back_broken) {
            // Read about this issue here: http://benalman.com/news/2009/09/chrome-browser-history-buggine/
            // and here: http://supportforums.blackberry.com/t5/Web-Development/Bold-Browser-Broken-history-go-1/td-p/107541
            ok(true, 'back button support inconsistent in chrome / blackberry (skip test)');
          } else {
            same(arr, ['', 'b', 'a'], 'back button and history.go(-1) should work');
          }

          $(window).unbind('hashchange');
          var events = $.data(window, 'events');
          ok(!events || !events.hashchange, 'hashchange event unbound');
        },

        ['all_done']

      );

    });

  });

})(jQuery); // END CLOSURE

jquery.ba-hashchange.js
/*!
 * jQuery hashchange event - v2.0 - 4/18/2020
 * http://benalman.com/projects/jquery-hashchange-plugin/
 *
 * Copyright (c) 2010 "Cowboy" Ben Alman
 * Dual licensed under the MIT and GPL licenses.
 * http://benalman.com/about/license/
 */

// Script: jQuery hashchange event
//
// *Version: 2.0, Last updated: 4/18/2020*
//
// Project Home - http://benalman.com/projects/jquery-hashchange-plugin/
// GitHub       - http://github.com/cowboy/jquery-hashchange/
// Source       - http://github.com/cowboy/jquery-hashchange/raw/master/jquery.ba-hashchange.js
// (Minified)   - http://github.com/cowboy/jquery-hashchange/raw/master/jquery.ba-hashchange.min.js (0.8kb gzipped)
//
// About: License
//
// Copyright (c) 2010 "Cowboy" Ben Alman,
// Dual licensed under the MIT and GPL licenses.
// http://benalman.com/about/license/
//
// About: Examples
//
// These working examples, complete with fully commented code, illustrate a few
// ways in which this plugin can be used.
//
// hashchange event - http://benalman.com/code/projects/jquery-hashchange/examples/hashchange/
// document.domain - http://benalman.com/code/projects/jquery-hashchange/examples/document_domain/
//
// About: Support and Testing
//
// Information about what version or versions of jQuery this plugin has been
// tested with, what browsers it has been tested in, and where the unit tests
// reside (so you can test it yourself).
//
// jQuery Versions - 1.2.6, 1.3.2, 1.4.1, 1.4.2
// Browsers Tested - Internet Explorer 6-8, Firefox 2-4, Chrome 5-6, Safari 3.2-5,
//                   Opera 9.6-10.60, iPhone 3.1, Android 1.6-2.2, BlackBerry 4.6-5.
// Unit Tests      - http://benalman.com/code/projects/jquery-hashchange/unit/
//
// About: Known issues
//
// While this jQuery hashchange event implementation is quite stable and
// robust, there are a few unfortunate browser bugs surrounding expected
// hashchange event-based behaviors, independent of any JavaScript
// window.onhashchange abstraction. See the following examples for more
// information:
//
// Chrome: Back Button - http://benalman.com/code/projects/jquery-hashchange/examples/bug-chrome-back-button/
// Firefox: Remote XMLHttpRequest - http://benalman.com/code/projects/jquery-hashchange/examples/bug-firefox-remote-xhr/
// WebKit: Back Button in an Iframe - http://benalman.com/code/projects/jquery-hashchange/examples/bug-webkit-hash-iframe/
// Safari: Back Button from a different domain - http://benalman.com/code/projects/jquery-hashchange/examples/bug-safari-back-from-diff-domain/
//
// Also note that should a browser natively support the window.onhashchange
// event, but not report that it does, the fallback polling loop will be used.
//

(function ($, window, undefined) {
  '$:nomunge'; // Used by YUI compressor.

  // Reused string.
  var str_hashchange = 'hashchange',

    // Method / object references.
    doc = document,
    fake_onhashchange,
    special = $.event.special,

    // Does the browser support window.onhashchange? Note that IE8 running in
    // IE7 compatibility mode reports true for 'onhashchange' in window, even
    // though the event isn't supported, so also test document.documentMode.
    doc_mode = doc.documentMode,
    supports_onhashchange = 'on' + str_hashchange in window && (doc_mode === undefined || doc_mode > 7);

  // Get location.hash (or what you'd expect location.hash to be) sans any
  // leading #. Thanks for making this necessary, Firefox!
  function get_fragment(url) {
    url = url || location.href;
    return '#' + url.replace(/^[^#]*#?(.*)$/, '$1');
  };

  // Method: jQuery.fn.hashchange
  //
  // Bind a handler to the window.onhashchange event or trigger all bound
  // window.onhashchange event handlers. This behavior is consistent with
  // jQuery's built-in event handlers.
  //
  // Usage:
  //
  // > jQuery(window).hashchange( [ handler ] );
  //
  // Arguments:
  //
  //  handler - (Function) Optional handler to be bound to the hashchange
  //    event. This is a "shortcut" for the more verbose form:
  //    jQuery(window).bind( 'hashchange', handler ). If handler is omitted,
  //    all bound window.onhashchange event handlers will be triggered. This
  //    is a shortcut for the more verbose
  //    jQuery(window).trigger( 'hashchange' ). These forms are described in
  //    the <hashchange event> section.
  //
  // Returns:
  //
  //  (jQuery) The initial jQuery collection of elements.

  // Allow the "shortcut" format $(elem).hashchange( fn ) for binding and
  // $(elem).hashchange() for triggering, like jQuery does for built-in events.
  $.fn[str_hashchange] = function (fn) {
    return fn ? this.bind(str_hashchange, fn) : this.trigger(str_hashchange);
  };

  // Property: jQuery.fn.hashchange.delay
  //
  // The numeric interval (in milliseconds) at which the <hashchange event>
  // polling loop executes. Defaults to 50.

  // Property: jQuery.fn.hashchange.domain
  //
  // If you're setting document.domain in your JavaScript, and you want hash
  // history to work in IE6/7, not only must this property be set, but you must
  // also set document.domain BEFORE jQuery is loaded into the page. This
  // property is only applicable if you are supporting IE6/7 (or IE8 operating
  // in "IE7 compatibility" mode).
  //
  // In addition, the <jQuery.fn.hashchange.src> property must be set to the
  // path of the included "document-domain.html" file, which can be renamed or
  // modified if necessary (note that the document.domain specified must be the
  // same in both your main JavaScript as well as in this file).
  //
  // Usage:
  //
  // jQuery.fn.hashchange.domain = document.domain;

  // Property: jQuery.fn.hashchange.src
  //
  // If, for some reason, you need to specify an Iframe src file (for example,
  // when setting document.domain as in <jQuery.fn.hashchange.domain>), you can
  // do so using this property. Note that when using this property, history
  // won't be recorded in IE6/7 until the Iframe src file loads. This property
  // is only applicable if you are supporting IE6/7 (or IE8 operating in "IE7
  // compatibility" mode).
  //
  // Usage:
  //
  // jQuery.fn.hashchange.src = 'path/to/file.html';

  $.fn[str_hashchange].delay = 50;
  /*
  $.fn[ str_hashchange ].domain = null;
  $.fn[ str_hashchange ].src = null;
  */

  // Event: hashchange event
  //
  // Fired when location.hash changes. In browsers that support it, the native
  // HTML5 window.onhashchange event is used, otherwise a polling loop is
  // initialized, running every <jQuery.fn.hashchange.delay> milliseconds to
  // see if the hash has changed. In IE6/7 (and IE8 operating in "IE7
  // compatibility" mode), a hidden Iframe is created to allow the back button
  // and hash-based history to work.
  //
  // Usage as described in <jQuery.fn.hashchange>:
  //
  // > // Bind an event handler.
  // > jQuery(window).hashchange( function(e) {
  // >   var hash = location.hash;
  // >   ...
  // > });
  // >
  // > // Manually trigger the event handler.
  // > jQuery(window).hashchange();
  //
  // A more verbose usage that allows for event namespacing:
  //
  // > // Bind an event handler.
  // > jQuery(window).bind( 'hashchange', function(e) {
  // >   var hash = location.hash;
  // >   ...
  // > });
  // >
  // > // Manually trigger the event handler.
  // > jQuery(window).trigger( 'hashchange' );
  //
  // Additional Notes:
  //
  // * The polling loop and Iframe are not created until at least one handler
  //   is actually bound to the 'hashchange' event.
  // * If you need the bound handler(s) to execute immediately, in cases where
  //   a location.hash exists on page load, via bookmark or page refresh for
  //   example, use jQuery(window).hashchange() or the more verbose
  //   jQuery(window).trigger( 'hashchange' ).
  // * The event can be bound before DOM ready, but since it won't be usable
  //   before then in IE6/7 (due to the necessary Iframe), recommended usage is
  //   to bind it inside a DOM ready handler.

  // Override existing $.event.special.hashchange methods (allowing this plugin
  // to be defined after jQuery BBQ in BBQ's source code).
  special[str_hashchange] = $.extend(special[str_hashchange], {

    // Called only when the first 'hashchange' event is bound to window.
    setup: function () {
      // If window.onhashchange is supported natively, there's nothing to do..
      if (supports_onhashchange) { return false; }

      // Otherwise, we need to create our own. And we don't want to call this
      // until the user binds to the event, just in case they never do, since it
      // will create a polling loop and possibly even a hidden Iframe.
      $(fake_onhashchange.start);
    },

    // Called only when the last 'hashchange' event is unbound from window.
    teardown: function () {
      // If window.onhashchange is supported natively, there's nothing to do..
      if (supports_onhashchange) { return false; }

      // Otherwise, we need to stop ours (if possible).
      $(fake_onhashchange.stop);
    }

  });

  // fake_onhashchange does all the work of triggering the window.onhashchange
  // event for browsers that don't natively support it, including creating a
  // polling loop to watch for hash changes and in IE 6/7 creating a hidden
  // Iframe to enable back and forward.
  fake_onhashchange = (function () {
    var self = {},
      timeout_id,

      // Remember the initial hash so it doesn't get triggered immediately.
      last_hash = get_fragment(),

      fn_retval = function (val) { return val; },
      history_set = fn_retval,
      history_get = fn_retval;

    // Start the polling loop.
    self.start = function () {
      timeout_id || poll();
    };

    // Stop the polling loop.
    self.stop = function () {
      timeout_id && clearTimeout(timeout_id);
      timeout_id = undefined;
    };

    // This polling loop checks every $.fn.hashchange.delay milliseconds to see
    // if location.hash has changed, and triggers the 'hashchange' event on
    // window when necessary.
    function poll() {
      var hash = get_fragment(),
        history_hash = history_get(last_hash);

      if (hash !== last_hash) {
        history_set(last_hash = hash, history_hash);

        $(window).trigger(str_hashchange);

      } else if (history_hash !== last_hash) {
        location.href = location.href.replace(/#.*/, '') + history_hash;
      }

      timeout_id = setTimeout(poll, $.fn[str_hashchange].delay);
    };

    // vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
    // vvvvvvvvvvvvvvvvvvv REMOVE IF NOT SUPPORTING IE6/7/8 vvvvvvvvvvvvvvvvvvv
    // vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
    (window.navigator.userAgent.indexOf("MSIE ") > -1 || !!window.navigator.userAgent.match(/Trident.*rv\:11\./) || navigator.appName == 'Microsoft Internet Explorer') && window.attachEvent && !window.addEventListener && !supports_onhashchange && (function () {
      // Not only do IE6/7 need the "magical" Iframe treatment, but so does IE8
      // when running in "IE7 compatibility" mode.

      var iframe,
        iframe_src;

      // When the event is bound and polling starts in IE 6/7, create a hidden
      // Iframe for history handling.
      self.start = function () {
        if (!iframe) {
          iframe_src = $.fn[str_hashchange].src;
          iframe_src = iframe_src && iframe_src + get_fragment();

          // Create hidden Iframe. Attempt to make Iframe as hidden as possible
          // by using techniques from http://www.paciellogroup.com/blog/?p=604.
          iframe = $('<iframe tabindex="-1" title="empty"/>').hide()

            // When Iframe has completely loaded, initialize the history and
            // start polling.
            .one('load', function () {
              iframe_src || history_set(get_fragment());
              poll();
            })

            // Load Iframe src if specified, otherwise nothing.
            .attr('src', iframe_src || 'javascript:0')

            // Append Iframe after the end of the body to prevent unnecessary
            // initial page scrolling (yes, this works).
            .insertAfter('body')[0].contentWindow;

          // Whenever `document.title` changes, update the Iframe's title to
          // prettify the back/next history menu entries. Since IE sometimes
          // errors with "Unspecified error" the very first time this is set
          // (yes, very useful) wrap this with a try/catch block.
          doc.onpropertychange = function () {
            try {
              if (event.propertyName === 'title') {
                iframe.document.title = doc.title;
              }
            } catch (e) { }
          };

        }
      };

      // Override the "stop" method since an IE6/7 Iframe was created. Even
      // if there are no longer any bound event handlers, the polling loop
      // is still necessary for back/next to work at all!
      self.stop = fn_retval;

      // Get history by looking at the hidden Iframe's location.hash.
      history_get = function () {
        return get_fragment(iframe.location.href);
      };

      // Set a new history item by opening and then closing the Iframe
      // document, *then* setting its location.hash. If document.domain has
      // been set, update that as well.
      history_set = function (hash, history_hash) {
        var iframe_doc = iframe.document,
          domain = $.fn[str_hashchange].domain;

        if (hash !== history_hash) {
          // Update Iframe with any initial `document.title` that might be set.
          iframe_doc.title = doc.title;

          // Opening the Iframe's document after it has been closed is what
          // actually adds a history entry.
          iframe_doc.open();

          // Set document.domain for the Iframe document as well, if necessary.
          domain && iframe_doc.write('\x3cscript>document.domain="' + domain + '"\x3c/script>');

          iframe_doc.close();

          // Update the Iframe's hash, for great justice.
          iframe.location.hash = hash;
        }
      };

    })();
    // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    // ^^^^^^^^^^^^^^^^^^^ REMOVE IF NOT SUPPORTING IE6/7/8 ^^^^^^^^^^^^^^^^^^^
    // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    return self;
  })();

})(jQuery, this);
